src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -5,207 +5,197 @@
+            _messageStream = new MemoryStream();
+            _handler = new HttpClientHandler();
+            _client = new HttpClient(_handler, disposeHandler: true);
+            _contentHeaderValueCache = new Dictionary<string, string>();
 
-            PropertyInfo pi = typeof(HttpWebRequest).GetProperty("ContentLength");
-            if (pi != null)
-            {
-                ContentLengthMethodSetter = pi.GetSetMethod();
-            }
+            _effectiveHttpMethod = args.Method;
+            _requestUrl = args.RequestUri;
+
+            _requestMessage = new HttpRequestMessage(new HttpMethod(this.ActualMethod), _requestUrl);
 
-            pi = typeof(HttpWebRequest).GetProperty("UserAgent");
-            if (pi != null)
+            // Now set the headers.
+            foreach (var keyValue in args.Headers)
             {
-                UserAgentMethodSetter = pi.GetSetMethod();
+                this.SetHeader(keyValue.Key, keyValue.Value);
             }
         }
-#endif
 
         /// <summary>
-        /// Creates a new instance of HttpWebRequestMessage.
+        /// Returns the collection of request headers.
         /// </summary>
-        /// <param name="args">Arguments for creating the request message.</param>
-        [SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors", Justification = "SetHeader is a safe virtual method to be called from the constructor")]
-        public HttpWebRequestMessage(DataServiceClientRequestMessageArgs args)
-            : base(args.ActualMethod)
+        public override IEnumerable<KeyValuePair<string, string>> Headers
         {
-            Util.CheckArgumentNull(args, "args");
-            Debug.Assert(args.RequestUri.IsAbsoluteUri, "request uri is not absolute uri");
-            Debug.Assert(
-                args.RequestUri.Scheme.Equals("http", StringComparison.OrdinalIgnoreCase) ||
-                args.RequestUri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase),
-                "request uri is not for HTTP");
-            this.effectiveHttpMethod = args.Method;
-            this.requestUrl = args.RequestUri;
-
-            this.httpRequest = HttpWebRequestMessage.CreateRequest(this.ActualMethod, this.Url, args);
-
-            // Now set the headers.
-            foreach (var keyValue in args.Headers)
+            get
             {
-                this.SetHeader(keyValue.Key, keyValue.Value);
+                if (_requestMessage.Content != null)
+                {
+                    return HttpHeadersToStringDictionary(_requestMessage.Headers).Concat(HttpHeadersToStringDictionary(_requestMessage.Content.Headers));
+                }
+
+                return HttpHeadersToStringDictionary(_requestMessage.Headers).Concat(_contentHeaderValueCache);
             }
         }
 
-        #region Properties
-
-        /// <summary> Returns the request url. </summary>
+        /// <summary>
+        /// Gets or sets the request url.

```

Comment:
Is it right to say `gets or sets` if set will always throw an exception? Any reason then not to make it a Get only prop? Same question for the props following this that `throw new NotSupportedException();`

Reply:
these properties are overridden in this class. There are other places where they are set.